helix-core/src/test/java/org/apache/helix/integration/controller/TestTopologyMigration.java

```diff
@@ -0,0 +1,319 @@
+    _configAccessor.setClusterConfig(CLUSTER_NAME, clusterConfig);
+  }
+
+  /**
+   * Sets up initial resources and mock participants for the cluster.
+   */
+  private void setupInitResourcesAndParticipants() throws Exception {
+    for (int i = 0; i < RESOURCE_COUNT; i++) {
+      String dbName = "TestDB_" + i;
+
+      // Create and start participants for the resource
+      for (int j = 0; j < INSTANCES_PER_RESOURCE; j++) {
+        String participantName = "localhost_" + _nextStartPort;
+
+        InstanceConfig instanceConfig = new InstanceConfig.Builder().setDomain(
+                String.format("%s=%s, %s=%s", RACK, j % INIT_ZONE_COUNT, HOST, participantName))
+            .addTag(dbName).build(participantName);
+
+        _gSetupTool.getClusterManagementTool().addInstance(CLUSTER_NAME, instanceConfig);
+
+        MockParticipantManager participant = createParticipant(participantName);
+        participant.syncStart();
+        _nextStartPort++;
+        _participants.add(participant);
+      }
+
+      // Set up IdealState for the resource
+      IdealState is = createResourceWithWagedRebalance(CLUSTER_NAME, dbName,
+          BuiltInStateModelDefinitions.LeaderStandby.name(), PARTITIONS, REPLICA, REPLICA - 1);
+      is.setResourceGroupName(dbName);
+      _gSetupTool.getClusterManagementTool().setResourceIdealState(CLUSTER_NAME, dbName, is);
+
+      _allDBs.add(dbName);
+    }
+  }
+
+  /**
+   * Creates and starts a mock participant with a registered state model factory.
+   */
+  private MockParticipantManager createParticipant(String participantName) throws Exception {
+    MockParticipantManager participant =
+        new MockParticipantManager(ZK_ADDR, CLUSTER_NAME, participantName, 10, null);
+    participant.getStateMachineEngine()
+        .registerStateModelFactory("LeaderStandby", new LeaderStandbyStateModelFactory());
+    return participant;
+  }
+
+  /**
+   * Tests topology migration with and without domain updates, ensuring no shuffling occurs.
+   */
+  @Test
+  public void testTopologyMigrationByResourceGroup() {
+    Assert.assertTrue(_clusterVerifier.verifyByPolling());
+
+    // Step 1: Migrate to new topology in maintenance mode
+    Map<String, ExternalView> originalEVs = getEVs();
+    List<InstanceConfig> instanceConfigs =
+        _gSetupTool.getClusterManagementTool().getInstancesInCluster(CLUSTER_NAME).stream().map(
+            instanceName -> _gSetupTool.getClusterManagementTool()
+                .getInstanceConfig(CLUSTER_NAME, instanceName)).collect(Collectors.toList());
+
+    _gSetupTool.getClusterManagementTool().enableMaintenanceMode(CLUSTER_NAME, true);
+    setupClusterConfig(MIGRATED_TOPOLOGY, MZ);
+    migrateInstanceConfigTopology(instanceConfigs);
+    _gSetupTool.getClusterManagementTool().enableMaintenanceMode(CLUSTER_NAME, false);
+
+    // Verify cluster state after topology migration
+    Assert.assertTrue(_clusterVerifier.verifyByPolling());
```

Comment:
Same here. If the main thread move fast and check the moment controller did not change the mapping, it may pass this line right away.

Reply:
`verifyByPolling` will wait to verify for `.setWaitTillVerify(TestHelper.DEFAULT_REBALANCE_PROCESSING_WAIT_TIME)` which is 1500 milliseconds. 

I looked into other ways to see if controller has processed that MM is set and found no other indicator given that MAINTENANCE znode and maintenance history in CONTROLLER/HISTORY are both written by the client.

Shouldn't `verifyByPolling` wait long enough with the configured `waitTillVerify` to allow the controller to process MM?

Reply:
verifybypolling only checks IS == EV. If there is no difference between IS and EV then it will pass.

It has the possibility controller move slow and no IS / EV diff generated at the moment MM exited. Then it can move to next line immediately.

Reply:
I switched to using BestPossibleExternalViewVerifier that runs before cluster gets set to MM to ensure the cluster is converged with what verifier says is eventual BPS given the current state of the cluster and after we exit MM to ensure that cluster converges with the eventual BPS given the current state.

This will work because BestPossibleExternalViewVerifier actually computes BPS instead of just comparing IS and EV.